Changes based on Aaron Wang's original version

dongbohu opened the pull request:
This PR is for adage-server issue #294:
https://github.com/greenelab/adage-server/issues/294
It includes all changes that I made based on Aaron Wang's original `d3.network.js`.

Apparently Aaron wrote this file in a coding style that is different from the style that we have been using in adage-server. I tried to keep it as is.

It is a big PR. I wouldn't recommend either of you to spend too much time on it, because it is not the highest priority at this point.

mhuyck commented:
I started looking at this and haven't gotten very far yet but I have a question. I see that this is a new repository that is forked from Aaron's version. Are the changes you made here just the diffs between what Aaron has and what we are currently using for `adage-server`? In other words, if I do a `diff` between the post-merge version of what you have here, will the file match what we have at https://github.com/greenelab/adage-server/blob/master/interface/d3.network.js ?

dongbohu commented:
@mhuyck: Yes, this file is identical to `d3.network.js` in `adage-server/interface/` directory.

mhuyck commented on d3.network.js:
```diff
@@ -442,38 +575,101 @@ d3.network = function() {
+    function addGeneLegend(scale) {
+        if (!scale) {
+          scale = .8;
+        }
```
This is the same block of code used in `addEdgeLegend` to set a default value for `scale`. Does it make sense to create a `defaultLegendScale` constant to use in both places?

rzelayafavila replied:
:+1:

dongbohu replied:
I thought about this before. I didn't do it because the two bars are in different orientation: one is horizontal and the other is vertical. It may not always be a good idea to have exactly same default scale value.

dongbohu commented on d3.network.js:
```diff
@@ -87,8 +103,12 @@ d3.network = function() {
+          genes.sort(function(a,b){return b.query_degree/b.degree
+                                   -a.query_degree/a.degree;});
+        } else { // option in greenelab adage-server
+          genes.sort(function(a,b){return b.query_degree-a.query_degree;});
```
Hmm... all I remember is that I talked to @cgreene and he told me that it makes more sense to disable this option in adage-server. @cgreene, can you elaborate?

dongbohu replied:
Maybe we should add a new button on the UI to allow users to enable/disable the background correction option.

rzelayafavila replied:
@dongbohu I think creating a new button to allow users to enable/disable the background correction sounds like a good idea - not sure what users' opinions are on this though.

rzelayafavila replied:
Will we create an issue for creating this button in the future?

rzelayafavila commented on d3.network.js:
```diff
@@ -295,20 +353,84 @@ d3.network = function() {
+        return my;
+    };
+
+    // dhu: New function that enables legend bar of gene colors
```
I agree with this, more description in the comments would be helpful. I also agree that the 'New' adjective might be a bit misleading, since it will all stop being new right after this merge, but it is also helpful to know that it was not in the original `d3.network.js` file, so I am a bit torn about this part.

dongbohu replied:
As you guys have guessed, I wrote these comments initially to remind users (including myself) that these functions are not available in the original d3.network file. But since Aaron are not going to maintain d3.network.js any more, we never need to worry about it now. So it is a good idea to change them into more meaningful comments.

dongbohu commented on d3.network.js:
```diff
@@ -171,19 +201,43 @@ d3.network = function() {
+              selection.selectAll("circle.gene").each(function(g){
+                  g.fixed = true;
+              });
+            }
```
I agree. I think the best approach for now is to add a new option such as `enableForceSimulation` whose default is true, and define a function to enable/disable it, just like how `bgCorrection` is handled.

dongbohu commented on d3.network.js:
```diff
@@ -171,19 +201,43 @@ d3.network = function() {
+        }
+        else if (!geneColorEnabled && drawn) {
+            d3.select("svg.gene-legend").remove();
+        }
```
Agree. I will change them.

dongbohu commented on d3.network.js:
```diff
@@ -221,38 +275,42 @@ d3.network = function() {
 
-    my.filterEdgeWeight = function(min_edge_cut, max_edge_cut, cmpSign) {
+    my.filterWithWeightSign = function(min_edge_cut, max_edge_cut, node_cut,
+                                       weightSign) {
```
Makes sense. Thanks.

dongbohu commented on d3.network.js:
```diff
@@ -295,20 +353,84 @@ d3.network = function() {
+        if (!arguments.length) {
+          x = true;
+        }
+        geneColorEnabled = x;
```
It actually does track whether the gene circles should have colors based on gene expression value. When it is true, then the gene legend bar will be required to be drawn. (I do see your point of why it can also be called `geneLegendEnabled`). These two names mean exactly the same thing.

dongbohu commented on d3.network.js:
```diff
@@ -21,23 +21,39 @@ d3.network = function() {
     var r = function(d) {return d.query ? 20 : Math.max(10,5+ d.query_degree ? d.query_degree*10 : 10);};
     var w = function(d) {return Math.max(2, Math.abs(d.weight)*6);};
-    var edgeColor = d3.scale.linear().domain([min_edge, .15, max_edge])
+    var edgeColor = d3.scale.linear().domain([options.start_edge,.15,1])
```
Good catch! It should be `[min_edge, .15, max_edge]`.

dongbohu commented on d3.network.js:
```diff
@@ -21,23 +21,39 @@ d3.network = function() {
+        .range([options.start_color, options.mid_color, options.end_color]);
+
+    // dhu: gene circle's color
+    var geneColor = d3.scale.linear().domain([-1, .15, 1])
```
Yes, you are right. I will correct them.

dongbohu commented:
@mhuyck and @rzelayafavila: I addressed most of your comments. @mhuyck: for your comment on lines 216-223, I proposed a resolution. If you agree, I will implement it. Thanks.

rzelayafavila commented on d3.network.js:
```diff
@@ -216,8 +221,8 @@ d3.network = function() {
             // After layout, fix all nodes in place:
-            if (event.alpha * .99 < .005 && event.alpha > 0 ) {
-              selection.selectAll("circle.gene").each(function(g){
+            if (!forceSimulation && event.alpha * .99 < .005 && event.alpha > 0 ) {
```
Minor thing: there is a space between the `0` and the `)` here - since there is no space between the opening parenthesis and the beginning of the arguments (nor between the parentheses and the arguments in other parts of the code), I think it would probably be better to delete that space so it is just `0)`. What do you think?

dongbohu replied:
You are right, that extra space should be removed. Thanks.

rzelayafavila commented on d3.network.js:
```diff
@@ -37,6 +37,11 @@ d3.network = function() {
+
+        // Flag that indicates whether force simulation will be enabled
+        // when rendering the network. Default is true.
+        forceSimulation = true,
```
:+1:

rzelayafavila commented on d3.network.js:
```diff
@@ -512,6 +517,15 @@ d3.network = function() {
+          return forceSimulation;
+        }
+        forceSimulation = x;
+        return my;
```
Hmm - does whatever part of the code that calls this function know to expect either `forceSimulation` OR `my` to be the return value?

dongbohu replied:
Good question. I followed Aaron's style when writing this function. I wouldn't say that I am a big fan of such style though.

rzelayafavila replied:
Ah, I see

dongbohu commented:
@rzelayafavila: I just realized that in order to install `d3.network.js` from a github repo in `adage-server/interface/package.json`, d3-network repo requires its own `package.json`. So I just created a generic one with minimal info. Please take a look and let me know what you think. Thanks.

rzelayafavila commented on package.json:
```diff
@@ -0,0 +1,17 @@
+  "main": "d3.network.js",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/aaronkw/d3-network.git"
```
Hmm - are you sure this should have the location of Aaron's repo instead of this one? I don't quite follow, just because this is the package.json file for the repo in this package, with these modifications. Is it because we might want to make a PR back into Aaron's repo?

dongbohu replied:
Aaron told me a few days ago that he is not going to maintain this library any more. So he gave me the write permission. Once this PR is closed, I will sync grrene;lab/d3-network with his master version. So his repo is the `master of master`.

rzelayafavila replied:
I see, ok, wanted to make sure this was intentional.

rzelayafavila commented on package.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "d3-network",
+  "version": "0.5.0",
```
Just curious - why version 0.5.0?

dongbohu replied:
Aaron Wong's original package doesn't have a version number at all. I feel that it still needs some improvement before it can be called v1.0, so I named it `0.5.0`.
